src/gt4py/analysis/passes.py

```diff
@@ -530,20 +586,46 @@ def merge_with(self, candidate: "MultiStageMergingWrapper") -> None:
             else:
                 self._multi_stage.inputs[name] = extent
 
+    @property
+    def offset_check(self) -> Callable[[Dict[str, int]], bool]:
+        if self.k_offset_extends_domain:
+            check = lambda offset: any(offset.get(axis, 0) != 0 for axis in ("I", "J", "K"))
+        else:
+            check = lambda offset: any(offset.get(axis, 0) != 0 for axis in ("I", "J"))
+        return check
+
+    def has_cycle_with_parallel_axis_offset(self, graph: nx.DiGraph) -> bool:
+        """Check the field dependency graph for a cycle with non-zero offset."""
+        race_fields = []
+        for cycle in nx.simple_cycles(graph):
+            full_cycle = cycle + [cycle[0]]
+            for source, target in zip(full_cycle[:-1], full_cycle[1:]):
+                offsets = graph.get_edge_data(source, target).get("offsets", [])
+                if any(self.offset_check(offset) for offset in offsets):
+                    race_fields.append(cycle[0])
```

Comment:
You posted on slack that in FV3 you have a performance problem with this approach. The number of cycles can scale combinatorially. I have the following observation:

Effectively what you're doing here is just checking  if any edge that is in a cycle reads at an offset. there is absolutely no need to iterate over all cycles here, you're checking the same edges over and over again (they can be in multiple cycles).

I think there should be an at most O(V*E) algorithm to find the edges you want (e.g. starting a DFS at every node and see which paths lead back to the original node), maybe there is even an O(V+E) algorithm (at least that's the complexity of `networkx.algorithms.components.strongly_connected_components` i'm haven't convinced myself yet but that could be just what you need.)

Reply:
Also, this is run every time you try to merge, so you could improve on the runtime by instead of `race_fields.append` directly `return False`. Further, there might be a way to speed things up by using the knowledge that the edges within an existing multistage are fine

Reply:
So I have convinced myself now  and pushed all suggestions except the last point. Let me know if this solves the issue.

Reply:
@gronerl All great points: I should have immediately escaped instead of adding them to the list. Your changes look really good, and that unblocked the FV3 stencil compilation. Thanks again!